RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc

```diff
@@ -0,0 +1,586 @@
+      }
+      else
+      {
+         edm::LogError("PixelCPEClusterRepair")
+         << "@SUB = PixelCPEClusterRepair::localPosition"
+         << "Should never be here. PixelCPEClusterRepair should always be called with track angles. This is a bad error !!! ";
+         
+         theClusterParam.templXrec_ = theDetParam.theTopol->localX( theClusterParam.theCluster->x() ) - lorentz_drift * micronsToCm; // rough Lorentz drift correction
+         theClusterParam.templYrec_ = theDetParam.theTopol->localY( theClusterParam.theCluster->y() );
+      }
+   }   
+   else 
+   {
+      //--- Template Reco succeeded.
+      theClusterParam.hasFilledProb_ = true;
+
+      //--- Go from microns to centimeters
+      theClusterParam.templXrec_ *= micronsToCm;
+      theClusterParam.templYrec_ *= micronsToCm;
+      
+      //--- Go back to the module coordinate system
+      theClusterParam.templXrec_ += lp.x();
+      theClusterParam.templYrec_ += lp.y();
+   }
+   return;
+}
+
+
+
+
+//------------------------------------------------------------------
+//  localError() relies on localPosition() being called FIRST!!!
+//------------------------------------------------------------------
+LocalError
+PixelCPEClusterRepair::localError(DetParam const & theDetParam,  ClusterParam & theClusterParamBase) const
+{
+   
+   ClusterParamTemplate & theClusterParam = static_cast<ClusterParamTemplate &>(theClusterParamBase);
+   
+   //--- Default is the maximum error used for edge clusters.
+   //--- (never used, in fact: let comment it out, shut up the complains of the static analyzer, and save a few CPU cycles)
+   float xerr = 0.0f, yerr = 0.0f;
+   
+   // Check if the errors were already set at the clusters splitting level
+   if ( theClusterParam.theCluster->getSplitClusterErrorX() > 0.0f && theClusterParam.theCluster->getSplitClusterErrorX() < 7777.7f &&
+       theClusterParam.theCluster->getSplitClusterErrorY() > 0.0f && theClusterParam.theCluster->getSplitClusterErrorY() < 7777.7f )
+   {
+      xerr = theClusterParam.theCluster->getSplitClusterErrorX() * micronsToCm;
+      yerr = theClusterParam.theCluster->getSplitClusterErrorY() * micronsToCm;
+   }
+   else
+   {
+      //--- Check status of both template calls.
+      if ( (theClusterParam.ierr !=0) || (theClusterParam.ierr2 !=0) )   
+      {
+         // If reconstruction fails the hit position is calculated from cluster center of gravity
+         // corrected in x by average Lorentz drift. Assign huge errors.
+	 //
+         if(!GeomDetEnumerators::isTrackerPixel(theDetParam.thePart))
+            throw cms::Exception("PixelCPEClusterRepair::localPosition :")
+            << "A non-pixel detector type in here?";
+         
+         // Assign better errors based on the residuals for failed template cases
+         if ( GeomDetEnumerators::isBarrel(theDetParam.thePart) )
+         {
+            xerr = 55.0f * micronsToCm;
+            yerr = 36.0f * micronsToCm;
```

Comment:
are these magic numbers coming from somewhere new or can/should they be derived from existing constants?

Reply:
They are coming from a study around 2010-2011 done by Gavril Giurgiu, and are assigning the errors to the hit in the unlikely case when the template reco fails.  The 1D template reco almost never fails, so this is not really a problem, except in the code.

This can be redesigned a bit, since at this point (N years of data taking later) we actually have a much better how large these errors need to be.  We can also call the CPE Generic and get both the error and the position from there.

I made the comment but I am going to punt on this one.